### ISSUE TITLE
Dealing with RDF Bag

### ISSUE THREAD

**Issue by CiaranWelsh** [sys-bio/libOmexMeta#49]
The RDF `bag` construct is often used in SBML models. It would be better if libomexmeta was able to parse RDF `bag` and remove it without loss of information. 

Here are a couple of examples from BIOMD0000000385: 

```
                <rdf:Description rdf:about="#_272044">
                    <dc:creator>
                        <rdf:Bag>
                            <rdf:li rdf:parseType="Resource">
                                <vCard:N rdf:parseType="Resource">
                                    <vCard:Family>Chelliah</vCard:Family>
                                    <vCard:Given>Vijayalakshmi</vCard:Given>
                                </vCard:N>
                                <vCard:EMAIL>viji@ebi.ac.uk</vCard:EMAIL>
                                <vCard:ORG rdf:parseType="Resource">
                                    <vCard:Orgname>EMBL-EBI</vCard:Orgname>
                                </vCard:ORG>
                            </rdf:li>

                            <rdf:li rdf:parseType="Resource">
                                <vCard:N rdf:parseType="Resource">
                                    <vCard:Family>Nikoloski</vCard:Family>
                                    <vCard:Given>Zoran</vCard:Given>
                                </vCard:N>
                                <vCard:EMAIL>nikoloski@mpimp-golm.mpg.de</vCard:EMAIL>
                                <vCard:ORG rdf:parseType="Resource">
                                    <vCard:Orgname>Institute of Biochemistry and Biology, University of Potsdam, 14476
                                        Potsdam, Germany
                                    </vCard:Orgname>
                                </vCard:ORG>
                            </rdf:li>
                            <rdf:li rdf:parseType="Resource">
                                <vCard:N rdf:parseType="Resource">
                                    <vCard:Family>Arnold</vCard:Family>
                                    <vCard:Given>Anne</vCard:Given>
                                </vCard:N>
                                <vCard:EMAIL>arnold@mpimp-golm.mpg.de</vCard:EMAIL>
                                <vCard:ORG rdf:parseType="Resource">
                                    <vCard:Orgname>Max-Planck-Institute of Molecular Plant Physiology</vCard:Orgname>
                                </vCard:ORG>
                            </rdf:li>
                        </rdf:Bag>
                    </dc:creator>
                    <dcterms:created rdf:parseType="Resource">
                        <dcterms:W3CDTF>2011-10-19T14:51:13Z</dcterms:W3CDTF>
                    </dcterms:created>
                    <dcterms:modified rdf:parseType="Resource">
                        <dcterms:W3CDTF>2012-04-20T19:52:45Z</dcterms:W3CDTF>
                    </dcterms:modified>
                    <bqmodel:is>
                        <rdf:Bag>
                            <rdf:li rdf:resource="http://identifiers.org/biomodels.db/MODEL1109270001"/>
                        </rdf:Bag>
                    </bqmodel:is>
                    <bqmodel:is>
                        <rdf:Bag>
                            <rdf:li rdf:resource="http://identifiers.org/biomodels.db/BIOMD0000000385"/>
                        </rdf:Bag>
                    </bqmodel:is>
                    <bqmodel:isDescribedBy>
                        <rdf:Bag>
                            <rdf:li rdf:resource="http://identifiers.org/pubmed/22001849"/>
                        </rdf:Bag>
                    </bqmodel:isDescribedBy>
                    <bqmodel:is>
                        <rdf:Bag>
                            <rdf:li rdf:resource="http://identifiers.org/obo.go/GO:0019253"/>
                        </rdf:Bag>
                    </bqmodel:is>
                    <bqbiol:hasTaxon>
                        <rdf:Bag>
                            <rdf:li rdf:resource="http://identifiers.org/taxonomy/33090"/>
                        </rdf:Bag>
                    </bqbiol:hasTaxon>
                </rdf:Description>

```
Note the VCard is a separate issue that needs addressing (see #50) , aka converting to an Orcid ID construct. The rest however is easier to deal with and should look (approximately) like the following after parsing by libomexmeta: 

```
<http://omex-library.org/NewOmex.omex/NewModel.xml#_272044>  
    dc:created [
        dc:W3CDTF 2011-10-19T14:51:13Z
    ] .
    dc:modified [
        dc:W3CDTF 2011-10-19T14:51:13Z
    ] .
    bqmodel:is <http://identifiers.org/biomodels.db/MODEL1109270001> .
    bqmodel:is <http://identifiers.org/biomodels.db/BIOMD0000000385> .
    isDescribedBy <http://identifiers.org/pubmed/22001849> .
    bqmodel:is <http://identifiers.org/obo.go/GO:0019253> .
    bqbiol:hasTaxon <http://identifiers.org/taxonomy/33090> .

```

**Comment by CiaranWelsh** [comment]
General strategy - 
    1) iterate over xml nodes with libxml2.
    2) If we find a `RDF::bag` tag 
        a) one parent up is the predicate `p`
        b) two parents up is the subject `s`
        c) one child down is a list of resources `r_i` 
    3) the list of triples extracted from a bag then becomes `p` `s` `r_i` for all `i` in resources
    4) Remove the RDF bag construct from the local sbml. 
        a) the sbml string passed to `Editor` is a copy. We need to modify this in place to remove a bag element after annotations have been extracted. Otherwise they'll get added again

Easiest way to add this is to have another method in SBMLSemantic extraction and lump the operation in with the other SBML extraction operations.

**Comment by CiaranWelsh** [comment]
On second thoughts, it might be easier to process the RDF graph, rather than processing the sbml? In this option:

1. Read rdf as normal
2. sparql query for annotations containing rdf:bag
3. convert search results from 2 to singular annotations and add them to the graph
4. remove annotations found in 2

This clearly depends on being able to write a sparql query to match RDF bag elements. For instance, given 
```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix bqbiol: <http://biomodels.net/biology-qualifiers/> .
@prefix OMEXlib: <http://omex-library.org/> .
@prefix local: <http://omex-library.org/AnAwesomeOmex.omex/Model1.rdf#> .

<http://omex-library.org/NewOmex.omex/NewModel.xml#_1>
    bqbiol:is [
        rdf:_1 <http://identifiers.org/uniprot/P00742> ;
        a rdf:Bag
    ] .

```
How do we write such a query?

**Comment by CiaranWelsh** [comment]
This issue is solved by #59 . However, before we can actually integrate these changes into the main code base, we need to solve the VCard problem  #50

**Comment by CiaranWelsh** [comment]
resolved with #62
